heron/instance/src/java/com/twitter/heron/instance/HeronInstance.java

```diff
@@ -233,6 +233,15 @@ public void uncaughtException(Thread thread, Throwable exception) {
 
     // The actual uncaught exceptions handing logic
     private void handleException(Thread thread, Throwable exception) {
+      // We would fail fast when errors occur to avoid unexpected bad situations
+      if (exception instanceof Error) {
+        LOG.log(Level.SEVERE,
+            "Error caught in thread: " + thread.getName()
+                + " with thread id: " + thread.getId() + ". Process halting...",
+            exception);
+        Runtime.getRuntime().halt(1);
```

Comment:
I would prefer to halt() directly here, since:
1. It is more clear to know the process will fail fast without tracing other parts of the code.
2. It reduces extra operations, which are not guaranteed to execute during Error. For instance, catching exceptions will use memory, and it might not work if we encounter `java.lang.OutOfMemoryError: Java heap space`